### ISSUE TITLE
Modified version of PPI.py (batch_size=8) shows (random) spikes in loss, accuracy

### ISSUE THREAD

**Issue by chrisba73** [rusty1s/pytorch_geometric#1478]
## 🐛 Bug
I'm a first-time Pytorch Geometric user.  I modified PPI.py for a node classification problem that wanted to try, and I noticed some anomalies in the training/validation results, including "spikes" in loss and accuracy.

So, I came back to examples/PPI.py and modified it to try large batch sizes, and I can reproduce similar behavior on PPI, too.

## To Reproduce
Here is the modified version of examples/PPI.py.  I decreased the size of the parameters in the GAT layer, so that I could increase the batch size to 8.  I added initial_seed control to try and enhance predictability, but I don't think it matters.
```
import os.path as osp
import matplotlib.pyplot as plt 
import numpy as np
import torch
import torch.nn.functional as F
from torch_geometric.datasets import PPI
from torch_geometric.data import DataLoader
from torch_geometric.nn import GATConv
from sklearn.metrics import f1_score

path = osp.join(osp.dirname(osp.realpath(__file__)), '..', 'data', 'PPI')
train_dataset = PPI(path, split='train')
val_dataset = PPI(path, split='val')
test_dataset = PPI(path, split='test')
if 1:
    bs       = 8  # was: trn=1, val/test=2
    nheads   = 2  # was: 4
    nheads_f = 5  # was: 6
    nhidden  = 64 # was: 256
    Nepochs = 500 # was: 100

initial_seed = 100
torch.manual_seed(initial_seed)
np.random.seed(initial_seed)

train_loader = DataLoader(train_dataset, batch_size=bs, shuffle=True) 
val_loader = DataLoader(val_dataset, batch_size=bs, shuffle=False) 
test_loader = DataLoader(test_dataset, batch_size=bs, shuffle=False)

class Net(torch.nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = GATConv(train_dataset.num_features, nhidden, heads=nheads)
        self.lin1 = torch.nn.Linear(train_dataset.num_features, nheads * nhidden)
        self.conv2 = GATConv(nheads * nhidden, nhidden, heads=nheads)
        self.lin2 = torch.nn.Linear(nheads * nhidden, nheads * nhidden)
        self.conv3 = GATConv(nheads * nhidden, train_dataset.num_classes, heads=nheads_f, concat=False)
        self.lin3 = torch.nn.Linear(nheads * nhidden, train_dataset.num_classes)

    def forward(self, x, edge_index):
        x = F.elu(self.conv1(x, edge_index) + self.lin1(x))
        x = F.elu(self.conv2(x, edge_index) + self.lin2(x))
        x = self.conv3(x, edge_index) + self.lin3(x)
        return x

device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
model = Net().to(device)
loss_op = torch.nn.BCEWithLogitsLoss()
optimizer = torch.optim.Adam(model.parameters(), lr=0.005)

def train():
    model.train()

    total_loss = 0
    for data in train_loader:
        data = data.to(device)
        optimizer.zero_grad()
        preds = model(data.x, data.edge_index)
        loss = loss_op(preds, data.y)
        total_loss += loss.item() * data.num_graphs
        loss.backward()
        optimizer.step()
    return total_loss / len(train_loader.dataset)


@torch.no_grad()
def test(loader):
    model.eval()

    ys, preds = [], []
    for data in loader:
        ys.append(data.y)
        out = model(data.x.to(device), data.edge_index.to(device))
        preds.append((out > 0).float().cpu())

    y, pred = torch.cat(ys, dim=0).numpy(), torch.cat(preds, dim=0).numpy()
    return f1_score(y, pred, average='micro') if pred.sum() > 0 else 0

def plot_training(Res, title):
    fig, ax = plt.subplots(figsize=(8, 6))
    plt.plot(Res[:,0], Res[:,1], label='Loss', linewidth=2)
    plt.plot(Res[:,0], Res[:,2], label='Val F1', linewidth=2)
    plt.plot(Res[:,0], Res[:,3], label='Test F1', linewidth=2)
    plt.legend()
    ax.set_ylim(0, 1); ax.grid(True); plt.title(title); plt.show()

Res = np.zeros((Nepochs,4))
for epoch in range(1, Nepochs+1):
    loss = train()
    val_f1 = test(val_loader)
    test_f1 = test(test_loader)
    print('Epoch: {:02d}, Loss: {:.4f}, Val: {:.4f}, Test: {:.4f}'.format(
        epoch, loss, val_f1, test_f1))
    Res[epoch-1,:] = np.array([epoch, loss, val_f1, test_f1])
plot_training(Res, 'PPI: BS=%d, nheads=%d/%d, nhidden=%d, seed=%d' % (bs, nheads, nheads_f, nhidden, initial_seed))

```
## Expected behavior

I would expect that the training/val/test accuracy curves would be relatively smooth, without significant dropout even if the training batch size is > 1.  Here are 4 plots showing the random spikes, 3 from my Windows machine, one from my Linux machine:

**Windows**:
![bs8_nheads2,5_nh64_seed100_run#1](https://user-images.githubusercontent.com/68907553/88677739-a8f03e00-d0bb-11ea-84b7-7d7985d6c133.png)
![bs8_nheads2,5_nh64_seed100_run#2](https://user-images.githubusercontent.com/68907553/88677750-ab529800-d0bb-11ea-8b3b-e2d1f81329ac.png)
![bs8_nheads2,5_nh64_seed100_run#3](https://user-images.githubusercontent.com/68907553/88677757-ad1c5b80-d0bb-11ea-8f42-b62b09549df7.png)

**Linux**:
![bs8_nheads2,5_nh64_seed100_run#4](https://user-images.githubusercontent.com/68907553/88677762-ae4d8880-d0bb-11ea-82c3-7da81c2b28d2.png)

## Environment

Environment 1:
* OS: Windows 10
* Python version: 3.8
* PyTorch version:  1.5.1
* PyTorch Geometric version:  1.6.0 (pip install)
* CUDA/cuDNN version: 10.1
* GCC version: N/A

Environment 2:
* OS: Ubuntu 20.04
* Python version: 3.8
* PyTorch version:  1.4
* PyTorch Geometric version:  1.6.0 (pip install)
* CUDA/cuDNN version: 10.1
* GCC version: 8.4

## Additional context
Let me know if there is additional information I can supply, to help clarify or debug the issue.

**Comment by chrisba73** [comment]
I decided to try (on Ubuntu) the same parameters as those in the original examples/ppi.py, but running for many iterations.  See the chart below; I'm not looking at the long-term trends (e.g. due to overtraining), but the spikes.  So, the problem may not be strictly related to large batch size.

```
bs       = 1
nheads   = 4
nheads_f = 6
nhidden  = 256
Nepochs = 3000
```
![Figure_2](https://user-images.githubusercontent.com/68907553/88705861-1a40e880-d0de-11ea-8f77-15cf966c629a.png)

**Comment by chrisba73** [comment]
I decided to try DGL (I'm new to DGL, too).  I saw that there was a similar example train_ppi.py, in [dgl/examples/pytorch/gat](https://github.com/dmlc/dgl/tree/master/examples/pytorch/gat).  I modified it to not stop early (no "patience"), to always compute train/val/test, and to plot results.  Here is what I got, first run:

**PPI, DGL, 400 iterations**
![PPI_DGL](https://user-images.githubusercontent.com/68907553/88723868-9d226d00-d0f7-11ea-8697-18dea9a7f724.jpg)

I'm confused.  My current view is:
* This is not a dataset problem; I saw this on my own synthetic dataset first, then went back to PPI and recreated it.
* This doesn't look like a framework issue; similar behavior occurs on Pytorch Geometric and DGL.
* This is not a batch-size issue.
* This is somehow related to sensitivity of the matrices or attention vector within GAT itself, to a slight weight update (what else is left).

If you can recommend another node-classification example to try, I could do some more exploration.  My starting point had been the recommendation from #1286.

**Comment by rusty1s** [comment]
Hi and thanks for the experimental study! It‘s true that GNNs may encounter some instabilities while training, but as far as I know, there is no thorough study about this. Seems like the loss landscape has some crazy hills in it :P This also holds for other GNNs. As far as I know, the best way to reduce these instabilities is to apply gradient norm before updating the weights.

**Comment by chrisba73** [comment]
OK, thanks @rusty1s.  I tried 
```
        loss.backward()     
        torch.nn.utils.clip_grad_norm_(model.parameters(), clip, math.inf) 
        optimizer.step()
```
last night, with mixed results.  It limits the amplitude of spikes and deviations in loss/accuracy, but doesn't get rid of them, for what seems to be a reasonable value of `clip`.  Also, I've seen a couple of instances where with clipping: instead of there being a spike in the loss function, there is a 'boxcar', that can last longer (maybe 30-40 epochs).  This might make things worse, depending on spike frequency.

Gradient clipping seems to be an available tool that's rarely used.  Doing a web search turns up little; I've seen it mentioned in very few tech papers.  So, I'm not sure about relying on this, vs. understanding if there is something more fundamental going on with the GAT structure, PPI data, and/or graph-based back-propagation (e.g.)

Also, for my true application, data augmentation (of portions of the node feature vectors) is quite relevant, and something that I want to use.  But, I noticed before that this tends to produce even more spikes.

Please let me know if you have any other thoughts.  I'm not sure what to do with this as an issue.  I might try and histogram gradients at each epoch, for each of model.parameters(), and estimate something related to outliers, but I'd need support to understand what it means, etc.

**Comment by chrisba73** [comment]
An update on this:  I've continued to see spiky behavior on several other PytGeo models that I've been working on.  I use gradient clipping on all of these, generally finding limited success in fighting the spikes.

On another PytGeo problem that I was working on today, I was using SGD.  I switched to Adam to get better performance (and did), but that brought in the spiky behavior in the loss and accuracy.  With this "hint", I tried using the Adamax optimizer instead of the Adam optimizer, in the PPI example above.  This addressed the spike problem.  Performance is different, though:
  Adam: Epoch: 500, Loss: 0.1020, Val: 0.9119, Test: 0.9347
  Adamax: Epoch: 500, Loss: 0.1531, Val: 0.8798, Test: 0.9018

Googling this, there is some relevant discussion here:  https://stats.stackexchange.com/questions/303857/explanation-of-spikes-in-training-loss-vs-iterations-with-adam-optimizer
